bunsen/bunsen-avro/src/main/java/com/cerner/bunsen/avro/converters/DefinitionToAvroVisitor.java

```diff
@@ -555,8 +555,10 @@ public HapiConverter<Schema> visitReference(String elementName,
       List<String> referenceTypes,
       List<StructureField<HapiConverter<Schema>>> children) {
 
-    // Generate a record name based on the type of references it can contain.
-    String recordName = referenceTypes.stream().collect(Collectors.joining()) + "Reference";
+    // Generate a record name based on the element name.
+    String recordName = elementName + "Reference";
+    // Remove extension type [x] as this creates issue while identifying schema.
```

Comment:
Hmm, where do we depend on this `recordName` other than using it in the keys for the `visitedConverters` cache? If it is just the cache, can you expand on how this name was causing issues? It seems to me it should work as long as the store and look-up in the cache use the same key.

Reply:
So, in case of Patient schema, it has `identifier` at its root element which has `assigner` as its sub-element. Now, when we are visiting reference for `Patient.identifier.assigner`, the elementName comes as `Organization` making the recordName as `OrganizationReference`  and for which schema does not allow identifier anymore because we have limited the maxDepth to 1. This schema for OrganizationReference gets cached and its fed to `managingOrganization` reference since its element name comes out to be `Organization` again making record name `OrganizationReference` again.

Reply:
Yeah, that part I understand and I think that is the root issue you are trying to fix. But I am wondering why these `[x]` parts were causing a problem. There might be something that I am missing here because this is related to the other comment/error too, but it seems to me that we are using this `recordName` just for caching converters in `visitedConverters` so any name should work. Can you confirm whether this understanding is correct or not? If yes, why does having `[x]` in the name matter?

Reply:
On the same lines, when we try to print schema, Schema class has a method `validateName` which checks the name and flags if any of the character in the name is not digit or alphabet or underscore:

`for (int i = 1; i < length; i++) {
      char c = name.charAt(i);
      if (!(Character.isLetterOrDigit(c) || c == '_'))
        throw new SchemaParseException("Illegal character in: " + name);
    }`

Reply:
I see, please consider updating the comment with a reference to name "validation"; and thanks for the clarification.